### ISSUE TITLE
MongoDB: duplicates in transaction_traces collection

### ISSUE THREAD

**Issue by Jeiwan** [EOSIO/eos#4985]
At some point the plugin started to produce duplicates in `transaction_traces` collection. For every transaction there are two records in `transaction_traces` and they're created almost immediately (less than a second between them). What's more important is that the first record contains wrong `global_sequence` numbers for its actions – they're from the previous block.

Example:
```
> db.transaction_traces.find({id:"7df71d8bd04407113346aa4dd42e68b56426acd8f10cf393f132a1b72cb1496b"}).pretty()

{
	"_id" : ObjectId("5b59d9ec99298f0001e1d13a"),
	"id" : "7df71d8bd04407113346aa4dd42e68b56426acd8f10cf393f132a1b72cb1496b",
	"receipt" : {
		"status" : "executed",
		"cpu_usage_us" : 154,
		"net_usage_words" : 13
	},
	"elapsed" : 154,
	"net_usage" : 104,
	"scheduled" : false,
	"action_traces" : [
		{
			"receipt" : {
				"receiver" : "blocktwitter",
				"act_digest" : "53e4a9911ac44c82ae11d5e1dcb188a1fce6aa6de69bd5bd4d0db1fabc10d39f",
				"global_sequence" : 98979615,
				"recv_sequence" : 75540645,
				"auth_sequence" : [
					[
						"chaintwitter",
						40293261
					]
				],
				"code_sequence" : 1,
				"abi_sequence" : 1
			},
			"act" : {
				"account" : "blocktwitter",
				"name" : "tweet",
				"authorization" : [
					{
						"actor" : "chaintwitter",
						"permission" : "active"
					}
				],
				"data" : "0a5745204c4f564520424d"
			},
			"elapsed" : 99,
			"cpu_usage" : 0,
			"console" : "",
			"total_cpu_usage" : 0,
			"trx_id" : "7df71d8bd04407113346aa4dd42e68b56426acd8f10cf393f132a1b72cb1496b",
			"inline_traces" : [ ]
		}
	],
	"except" : null,
	"createdAt" : ISODate("2018-07-26T14:25:48.233Z")
}
{
	"_id" : ObjectId("5b59d9ed99298f0001e1dce8"),
	"id" : "7df71d8bd04407113346aa4dd42e68b56426acd8f10cf393f132a1b72cb1496b",
	"receipt" : {
		"status" : "executed",
		"cpu_usage_us" : 347,
		"net_usage_words" : 13
	},
	"elapsed" : 239,
	"net_usage" : 104,
	"scheduled" : false,
	"action_traces" : [
		{
			"receipt" : {
				"receiver" : "blocktwitter",
				"act_digest" : "53e4a9911ac44c82ae11d5e1dcb188a1fce6aa6de69bd5bd4d0db1fabc10d39f",
				"global_sequence" : 98979821,
				"recv_sequence" : 75540851,
				"auth_sequence" : [
					[
						"chaintwitter",
						40293467
					]
				],
				"code_sequence" : 1,
				"abi_sequence" : 1
			},
			"act" : {
				"account" : "blocktwitter",
				"name" : "tweet",
				"authorization" : [
					{
						"actor" : "chaintwitter",
						"permission" : "active"
					}
				],
				"data" : "0a5745204c4f564520424d"
			},
			"elapsed" : 129,
			"cpu_usage" : 0,
			"console" : "",
			"total_cpu_usage" : 0,
			"trx_id" : "7df71d8bd04407113346aa4dd42e68b56426acd8f10cf393f132a1b72cb1496b",
			"inline_traces" : [ ]
		}
	],
	"except" : null,
	"createdAt" : ISODate("2018-07-26T14:25:49.182Z")
}
```

This transaction was created on 27 July, after I updated the node to version 1.1. I noticed that there are no duplicates of transactions created on an older version. The issue is actual for all the transactions created after that date.

**Comment by heifner** [comment]
While replaying you only get one `transaction_trace`. You get two `transaction_trace`s when your node is speculating a block. To avoid this "extra" trace use `read-mode = head` in your config.ini.

**Comment by Jeiwan** [comment]
@heifner Thank you!

**Comment by qq1032246642** [comment]
I use `read-mode = head` in my config.ini, but it still get two `transaction_traces`. 
add log in chain_plugin
`wlog("chain.get_read_mode()==chain::db_read_mode::HEAD : ${read}",("read",my->chain->get_read_mode()==chain::db_read_mode::HEAD));
`
![image](https://user-images.githubusercontent.com/14926995/43946004-8b8f02be-9cb6-11e8-9edc-18daf433fa18.png)
![image](https://user-images.githubusercontent.com/14926995/43945982-7724e0dc-9cb6-11e8-8e8e-30a641f80974.png)
@Jeiwan  @heifner

**Comment by heifner** [comment]
@qq1032246642 what plugins are you running? Are you processing RPC requests?

**Comment by qq1032246642** [comment]
@heifner I only run default plugin  and sql_db_plugin. This node only accpets message.

**Comment by qq1032246642** [comment]
In controller.cpp
Although execute `trx_context.undo()`, but it still emit` accepted_transaction` and `applied_transaction` signals. Should I move emit in `else`? @heifner 
```
// call the accept signal but only once for this transaction
            if (!trx->accepted) {
               emit( self.accepted_transaction, trx);
               trx->accepted = true;
            }

            emit(self.applied_transaction, trace);


            if ( read_mode != db_read_mode::SPECULATIVE && pending->_block_status == controller::block_status::incomplete ) {
               //this may happen automatically in destructor, but I prefere make it more explicit
               trx_context.undo();ilog("trx_context undo");
            } else {
               restore.cancel();
               trx_context.squash();
            }
```
After change
```
if ( read_mode != db_read_mode::SPECULATIVE && pending->_block_status == controller::block_status::incomplete ) {
               //this may happen automatically in destructor, but I prefere make it more explicit
               trx_context.undo();ilog("trx_context undo");
            } else {
               if (!trx->accepted) {
                 emit( self.accepted_transaction, trx);
                 trx->accepted = true;
               }

               emit(self.applied_transaction, trace);
               restore.cancel();
               trx_context.squash();
            }
```

**Comment by luoyhang003** [comment]
I also meet this problem.

**Comment by luoyhang003** [comment]
@firesWu Have you solved your problem? I also had the same problem.

**Comment by firesWu** [comment]
@luoyhang003  add `read-mode = read-only` in config.ini
